File: apps/api-spring/src/main/java/com/jiandou/api/task/TaskPersistenceMutation.java
package com.jiandou.api.task;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Aggregates all persistence rows produced by one task-side state transition.
 * Repository adapters can persist this mutation in a single transaction.
 */
public final class TaskPersistenceMutation {

    private String taskId = "";
    private TaskRecord task;
    private final List<Map<String, Object>> attempts = new ArrayList<>();
    private final List<Map<String, Object>> statusHistoryRows = new ArrayList<>();
    private final List<Map<String, Object>> traceRows = new ArrayList<>();
    private final List<Map<String, Object>> stageRunRows = new ArrayList<>();
    private final List<Map<String, Object>> modelCallRows = new ArrayList<>();
    private final List<Map<String, Object>> materialRows = new ArrayList<>();
    private final List<Map<String, Object>> resultRows = new ArrayList<>();
    private final List<Map<String, Object>> queueEventRows = new ArrayList<>();
    private final List<Map<String, Object>> workerInstanceRows = new ArrayList<>();

    public TaskPersistenceMutation task(TaskRecord value) {
        this.task = value;
        this.taskId = value == null || value.id == null ? "" : value.id;
        return this;
    }

    public TaskPersistenceMutation taskId(String value) {
        this.taskId = value == null ? "" : value.trim();
        return this;
    }

    public TaskPersistenceMutation addAttempt(Map<String, Object> value) {
        add(attempts, value);
        return this;
    }

    public TaskPersistenceMutation addStatusHistory(Map<String, Object> value) {
        add(statusHistoryRows, value);
        return this;
    }

    public TaskPersistenceMutation addTrace(Map<String, Object> value) {
        add(traceRows, value);
        return this;
    }

    public TaskPersistenceMutation addStageRun(Map<String, Object> value) {
        add(stageRunRows, value);
        return this;
    }

    public TaskPersistenceMutation addModelCall(Map<String, Object> value) {
        add(modelCallRows, value);
        return this;
    }

    public TaskPersistenceMutation addMaterial(Map<String, Object> value) {
        add(materialRows, value);
        return this;
    }

    public TaskPersistenceMutation addResult(Map<String, Object> value) {
        add(resultRows, value);
        return this;
    }

    public TaskPersistenceMutation addQueueEvent(Map<String, Object> value) {
        add(queueEventRows, value);
        return this;
    }

    public TaskPersistenceMutation addWorkerInstance(Map<String, Object> value) {
        add(workerInstanceRows, value);
        return this;
    }

    public TaskRecord task() {
        return task;
    }

    public String taskId() {
        return taskId;
    }

    public List<Map<String, Object>> attempts() {
        return attempts;
    }

    public List<Map<String, Object>> statusHistoryRows() {
        return statusHistoryRows;
    }

    public List<Map<String, Object>> traceRows() {
        return traceRows;
    }

    public List<Map<String, Object>> stageRunRows() {
        return stageRunRows;
    }

    public List<Map<String, Object>> modelCallRows() {
        return modelCallRows;
    }

    public List<Map<String, Object>> materialRows() {
        return materialRows;
    }

    public List<Map<String, Object>> resultRows() {
        return resultRows;
    }

    public List<Map<String, Object>> queueEventRows() {
        return queueEventRows;
    }

    public List<Map<String, Object>> workerInstanceRows() {
        return workerInstanceRows;
    }

    private void add(List<Map<String, Object>> target, Map<String, Object> value) {
        if (value != null && !value.isEmpty()) {
            target.add(value);
        }
    }
}

File: backend/infrastructure/task_persistence_mutation.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from backend.domain.task_record import TaskRecord


@dataclass
class TaskPersistenceMutation:
    """Aggregates all persistence rows produced by one task-side state transition.

    Repository adapters persist this mutation's contents in a single transaction.
    Mirrors the Java TaskPersistenceMutation builder pattern.
    """

    task_id: str = ""
    task: TaskRecord | None = None
    attempts: list[dict[str, Any]] = field(default_factory=list)
    status_history_rows: list[dict[str, Any]] = field(default_factory=list)
    trace_rows: list[dict[str, Any]] = field(default_factory=list)
    stage_run_rows: list[dict[str, Any]] = field(default_factory=list)
    model_call_rows: list[dict[str, Any]] = field(default_factory=list)
    request_log_rows: list[dict[str, Any]] = field(default_factory=list)
    material_rows: list[dict[str, Any]] = field(default_factory=list)
    result_rows: list[dict[str, Any]] = field(default_factory=list)
    queue_event_rows: list[dict[str, Any]] = field(default_factory=list)
    worker_instance_rows: list[dict[str, Any]] = field(default_factory=list)

    def set_task(self, value: TaskRecord | None) -> TaskPersistenceMutation:
        self.task = value
        self.task_id = "" if value is None else (value.id or "")
        return self

    def set_task_id(self, value: str) -> TaskPersistenceMutation:
        self.task_id = (value or "").strip()
        return self

    def add_attempt(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.attempts.append(value)
        return self

    def add_status_history(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.status_history_rows.append(value)
        return self

    def add_trace(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.trace_rows.append(value)
        return self

    def add_stage_run(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.stage_run_rows.append(value)
        return self

    def add_model_call(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.model_call_rows.append(value)
        return self

    def add_request_log(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.request_log_rows.append(value)
        return self

    def add_material(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.material_rows.append(value)
        return self

    def add_result(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.result_rows.append(value)
        return self

    def add_queue_event(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.queue_event_rows.append(value)
        return self

    def add_worker_instance(self, value: dict[str, Any]) -> TaskPersistenceMutation:
        if value:
            self.worker_instance_rows.append(value)
        return self
